### PR TITLE
a couple of fixes

### PR DESCRIPTION
**Pull request by maxim-ky**
Hi guys,

At first, thank you for such a great work. AMQP-CPP is nice in terms of architecture. it's also well-documented. And your work is especially valuable because AMQP-CPP is the only asynchronous AMQP c++ library.

But it seems you have not enough resources to support multiple platforms/compilers because I faced a lot of issues during compiling this one under GCC 4.7.3/cmake 2.8. If this toolset is considered to be outdated, then just don't merge 40ebb42.
Other commits are more important, mainly the undefined-behavior one.

Regards

### CODE REVIEW

File: src/connectionimpl.cpp [hunk 1]
```diff
@@ -123,7 +123,9 @@ uint64_t ConnectionImpl::parse(const Buffer &buffer)
         try
         {
             // try to recognize the frame
-            ReceivedFrame receivedFrame(ReducedBuffer(buffer, processed), _maxFrame);
```
Comment (Kojoley) [inline src/connectionimpl.cpp]: Why do you think that It is an undefined behaviour?

12.2/5 "Temporary objects":

> The temporary to which the reference is bound or the temporary that is the complete object to a subobject of which the temporary is bound persists for the lifetime of the reference except as specified below. A temporary bound to a reference member in a constructor’s ctor-initializer (12.6.2) persists until the constructor exits. **A temporary bound to a reference parameter in a function call (5.2.2) persists until the completion of the full expression containing the call.**

Comment (maxim-ky) [inline src/connectionimpl.cpp]: > Why do you think that It is an undefined behaviour?

The next statement is:
`
if (receivedFrame.complete())
`
ReducedBuffer is destroyed here because that **full expression containing the call** is already evaluated.
But ReceivedFrame::complete uses stored reference to the destroyed ReducedBuffer.
`
bool ReceivedFrame::complete() const
{
    return _buffer.size() >= _payloadSize + 8;
}
`

**Comment by martijnotto** [comment]
Thanks for the pull-request. I was wondering if there is a specific reason that you added all the virtual destructors to all the interface classes. As far as I can see we never destruct them in the library but I maye not have looked properly. Did you find instances where this is actually done?

If they really are necessary could you please add the docblocks to them. Just a simple

    /**
     *  Destructor
     */

will do.

File: src/connectionimpl.cpp (hunk 1 above)
Comment (Kojoley) [inline src/connectionimpl.cpp]: The _full expression containing the call_ is [here](https://github.com/CopernicaMarketingSoftware/AMQP-CPP/blob/f3343ad1cb64f5a58849fcf0ee3cc22dece2393c/src/connectionimpl.cpp#L126-150).
Look at example in the [C++ Standard](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3797.pdf) on page 254.

**Comment by maxim-ky** [comment]
> I was wondering if there is a specific reason that you added all the virtual destructors to all the interface classes.

The reason is this GCC warning: *a class has virtual functions but non-virtual destructor.*

> As far as I can see we never destruct them in the library but I maye not have looked properly.

You're right. Mostly it's the matter of some common code style. 
a Herb Sutter's quote:
> Guideline #4: A base class destructor should be either public and virtual, or protected and nonvirtual.

the original link: http://www.gotw.ca/publications/mill18.htm

Does that sound appropriate to you?

File: src/connectionimpl.cpp (hunk 1 above)
Comment (maxim-ky) [inline src/connectionimpl.cpp]: I'm sorry, but that example looks irrelevant to me (there's nothing about saving a reference).
C++ expressions can't consist of several statements (not willing to quote Standard for you, sorry again)

Actually, you could compile that code snippet with an existing compiler and see the output.
http://pastebin.com/UcLrCkvL

Comment (Kojoley) [inline src/connectionimpl.cpp]: You are making copy of object here `B bbb = B( A() );`. It is completely different situation and does not belong to what you have in your patch.

Comment (maxim-ky) [inline src/connectionimpl.cpp]: Actually, copying doesn't occur here. I used that syntax to avoid so called *most vexing parse*, but if you have some doubts, you could use this code snippet which completely resembles the original code.
http://pastebin.com/erL8GGXs

**Comment by martijnotto** [comment]
That sounds perfectly logical to me. Could you update your code to include a simple docblock for the destructors? I'd be happy to merge it then.

**Comment by maxim-ky** [comment]
> Could you update your code to include a simple docblock for the destructors?

It's done.

**Comment by martijnotto** [comment]
Looking good. Thanks for the help.
